include/flang/lower/OpBuilder.h

```diff
@@ -55,6 +61,107 @@ class SymMap {
   }
 };
 
+/// Helper class that can be inherited from in order to
+/// facilitate mlir::OpBuilder usage.
+class OpBuilderHandler {
+public:
+  OpBuilderHandler(mlir::OpBuilder &b, mlir::Location l) : builder{b}, loc{l} {}
+  template <typename T, typename... Args>
+  auto create(Args... args) {
+    return builder.create<T>(loc, std::forward<Args>(args)...);
+  }
+  mlir::Value getIntegerConstant(mlir::Type integerType, std::int64_t);
+
+protected:
+  mlir::OpBuilder &builder;
+  mlir::Location loc;
+};
+
+/// Facilitate lowering to fir::loop
+class LoopCreator : public OpBuilderHandler {
+public:
+  LoopCreator(mlir::OpBuilder &b, mlir::Location l) : OpBuilderHandler{b, l} {}
+  LoopCreator(OpBuilderHandler &b) : OpBuilderHandler{b} {}
+  // In genLoop functions, lb, ub, and count arguments must have integer types.
+  // They will be automatically converted the IndexType if needed.
+  // Build loop [lb, ub) with step "step".
+  using BodyGenerator = std::function<void(OpBuilderHandler &, mlir::Value)>;
+  void genLoop(mlir::Value lb, mlir::Value ub, mlir::Value step,
+               const BodyGenerator &bodyGenerator);
+  /// Build loop [lb, ub) with step 1.
+  void genLoop(mlir::Value lb, mlir::Value ub,
+               const BodyGenerator &bodyGenerator);
+  /// Build loop [0, count) with step 1.
+  void genLoop(mlir::Value count, const BodyGenerator &bodyGenerator);
+
+private:
+  mlir::Type getIndexType();
+  mlir::Value convertToIndexType(mlir::Value integer);
+};
+
+/// Facilitate lowering of CHARACTER operation
+class CharacterOpsCreator : public OpBuilderHandler {
+public:
+  CharacterOpsCreator(mlir::OpBuilder &b, mlir::Location l)
+      : OpBuilderHandler{b, l} {}
+  CharacterOpsCreator(OpBuilderHandler &b) : OpBuilderHandler{b} {}
+  /// Interchange format to avoid inserting unbox/embox everywhere while
+  /// evaluating character expressions.
+  struct CharValue {
+    fir::ReferenceType getReferenceType();
+    fir::CharacterType getCharacterType();
+
+    mlir::Value reference;
+    mlir::Value len;
+  };
+
+  /// Copy count first characters of src into dest.
+  void genCopy(CharValue &dest, CharValue &src, mlir::Value count);
+  /// Pad str(from:to) with blanks. If to <= from, no padding is done.
```

Comment:
to.len < from.len ?

Reply:
Ambiguous name choice, thanks for raising this. I will rename:
`from` -> `lower`
`to` -> `upper`
`upper` and `lower` are consistant with substring range bounds name in https://github.com/flang-compiler/f18/blob/9c7406324dc3768886ec8d2baf92b0d9395291be/lib/evaluate/variable.h#L333-L335

Reply:
Done and comment updated.